Build failure: mbedtls

### Nixpkgs version

- Stable (25.11)

### Steps to reproduce

failure on staging-25.11

### Can Hydra reproduce this build failure?

Hydra doesn’t try to build the package.

### Link to Hydra build job

_No response_

### Relevant log output

```console
error: Cannot build '/nix/store/paczfsp0rlzph19vm8qvxh4mh8grrdwi-mbedtls-3.6.5.drv'.
       Reason: builder failed with exit code 1.
       Output paths:
         /nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5
       Last 51 log lines:
       > Running phase: unpackPhase
       > unpacking source archive /nix/store/bk0pjl2wbqrzv2xi51sq3z457p8kxjsk-source
       > source root is source
       > Running phase: patchPhase
       > applying patch /nix/store/2lk3ha364aakq8w85nz13app4jfmdd13-100-fix-gcc14-build.patch
       > patching file library/common.h
       > Running phase: updateAutotoolsGnuConfigScriptsPhase
       > Running phase: configurePhase
       > cmake flags: -GNinja -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_LOCALEDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/share/doc/Mbed -DCMAKE_INSTALL_INFODIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/share/man -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/include -DCMAKE_INSTALL_SBINDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5/lib -DCMAKE_STRIP=/nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/strip -DCMAKE_RANLIB=/nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/ranlib -DCMAKE_AR=/nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/m9gwd78mm4d6xgiys0cnclwxd3l3dki1-mbedtls-3.6.5 -DUSE_SHARED_MBEDTLS_LIBRARY=on -DGEN_FILES=off -DCMAKE_C_FLAGS=-fzero-init-padding-bits=unions
       > CMake Deprecation Warning at CMakeLists.txt:21 (cmake_minimum_required):
       >   Compatibility with CMake < 3.10 will be removed from a future version of
       >   CMake.
       >
       >   Update the VERSION argument <min> value.  Or, use the <min>...<max> syntax
       >   to tell CMake that the project requires at least <min> but has been updated
       >   to work with policies introduced by <max> or earlier.
       >
       > 
       > -- The C compiler identification is GNU 14.3.0
       > -- Detecting C compiler ABI info
       > -- Detecting C compiler ABI info - failed
       > -- Check for working C compiler: /nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/gcc
       > -- Check for working C compiler: /nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/gcc - broken
       > CMake Error at /nix/store/i3590ryd16kpjr07m57n7b0gws0kd3nj-cmake-4.1.2/share/cmake-4.1/Modules/CMakeTestCCompiler.cmake:67 (message):
       >   The C compiler
       >
       >     "/nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/gcc"
       >
       >   is not able to compile a simple test program.
       >
       >   It fails with the following output:
       >
       >     Change Dir: '/build/source/build/CMakeFiles/CMakeScratch/TryCompile-6PiARC'
       >
       >     Run Build Command(s): /nix/store/na838sk0jgzc9h1ncjjra9cl297n405v-ninja-1.13.1/bin/ninja -v cmTC_d596b
       >     [1/2] /nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/gcc   -fzero-init-padding-bits=unions -o CMakeFiles/cmTC_d596b.dir/testCCompiler.c.o -c /build/source/build/CMakeFiles/CMakeScratch/TryCompile-6PiARC/testCCompiler.c
       >     FAILED: [code=1] CMakeFiles/cmTC_d596b.dir/testCCompiler.c.o
       >     /nix/store/ian5gigl7wb820snlri0pqcclszzb09w-gcc-wrapper-14.3.0/bin/gcc   -fzero-init-padding-bits=unions -o CMakeFiles/cmTC_d596b.dir/testCCompiler.c.o -c /build/source/build/CMakeFiles/CMakeScratch/TryCompile-6PiARC/testCCompiler.c
       >     gcc: error: unrecognized command-line option '-fzero-init-padding-bits=unions'
       >     ninja: build stopped: subcommand failed.
       >
       >
       >
       >
       >
       >   CMake will not be able to correctly generate this project.
       > Call Stack (most recent call first):
       >   CMakeLists.txt:46 (project)
       >
       > 
       > -- Configuring incomplete, errors occurred!
       For full logs, run:
         nix log /nix/store/paczfsp0rlzph19vm8qvxh4mh8grrdwi-mbedtls-3.6.5.drv
```

### Additional context

_No response_

### System metadata

x64-86

### Notify maintainers

@raphaelr 

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+build+failure%22).
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/mbedtls/generic.nix
+++ b/pkgs/development/libraries/mbedtls/generic.nix
@@ -55,7 +55,7 @@ stdenv.mkDerivation rec {
     # https://github.com/Mbed-TLS/mbedtls/releases/tag/v3.3.0 below "Requirement changes".
     "-DGEN_FILES=off"
   ]
-  ++ lib.optionals stdenv.cc.isGNU [
+  ++ lib.optionals (stdenv.cc.isGNU && lib.versionAtLeast stdenv.cc.version "15.0") [
     # mbedtls widely uses a pattern of starting unions with an
     # unsigned int dummy member, and then initializing those unions to
     # { 0 }.  The problem with this is that it only initializes that

